File: sdks/kotlin/src/main/kotlin/dev/orion/sdk/model/ModelResponse.kt
package dev.orion.sdk.model

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject

/** Normalized result returned by a [ModelAdapter]. */
internal data class ModelResponse(
    /** Assistant text, which can be empty when tools are requested. */
    public val content: String = "",
    /** Tool calls in provider order. */
    public val toolCalls: List<ToolCall> = emptyList(),
    /** Provider-neutral reason generation stopped. */
    public val finishReason: FinishReason = if (toolCalls.isEmpty()) FinishReason.STOP else FinishReason.TOOL_CALLS,
    /** Token accounting reported by the provider. */
    public val usage: Usage = Usage(),
    /** Opaque continuation state owned by the adapter. */
    public val providerState: JsonObject = buildJsonObject {},
)

File: sdks/javascript/src/model/model-response.ts
import type {JsonObject} from "./json.js";

/** Normalized reason a model stopped generating output. */
export type FinishReason = "stop" | "tool_calls" | "length" | "content_filter" | "other";

/** Normalized token consumption. */
export interface Usage {

  /** Number of input tokens consumed. */
  readonly inputTokens: number;

  /** Number of output tokens generated. */
  readonly outputTokens: number;

}

/** Provider-neutral tool invocation proposed by a model. */
export interface ToolCall {

  /** Provider-supplied call identifier. */
  readonly id: string;

  /** Registered application tool name. */
  readonly name: string;

  /** Object-shaped JSON arguments; the application validates schema constraints. */
  readonly arguments: JsonObject;

}

/** Complete normalized response returned by a model adapter. */
export interface ModelResponse {

  /** Assistant text, which may be empty when tools were requested. */
  readonly content: string;

  /** Tool calls in provider order. */
  readonly toolCalls: readonly ToolCall[];

  /** Normalized termination reason. */
  readonly finishReason: FinishReason;

  /** Token usage, with zeroes when the provider omits accounting. */
  readonly usage: Usage;

  /** Opaque continuation data retained for the provider adapter. */
  readonly providerState: JsonObject;

}
